### PR TITLE
Correlo spin matrix detector

### PR DESCRIPTION
**Pull request by mhelias**
Hi,
I would like the correlospinmatrix_detector to be included in the master branch.
Best wishes,
Moritz

### CODE REVIEW

**Comment by apeyser** [comment]
@mhelias: Can you @ mention a second or third reviewer to nominate them as reviewer? To wake them up that it's time to look at this, as abigailm has done?

**Comment by mhelias** [comment]
Hi Alex,

where do I nominate reviewers? Jakob (wrote correlation_matrix_dteector) and David (will use correlospinmatrix_detector) would be suitable candidates.
@Abigail: removed commented out code, improved documentation.

**Comment by apeyser** [comment]
@mhelias: just say @jakobj would be good to code review this, then he should get notified of this thread and will hopefully volunteer to help. I don't know David's github handle --- but just mention him @(whatever David's handle is).

**Comment by mhelias** [comment]
@ddahmen and @jakobj would be suitable candidates to review the code

**Comment by jakobj** [comment]
i got the notification and am happy to have a look at it.

File: examples/nest/correlospinmatrix_detector.sli [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+
+/* simple example of usage for the
+   correlospinmatrix detector
+/*
```
Comment (jakobj) [inline examples/nest/correlospinmatrix_detector.sli]: this should be */. breaks the example.

File: examples/nest/correlospinmatrix_detector.sli [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+sg3 << /spike_times [25.] >> SetStatus
+
+
+sg1 csd << /receptor_type 0 >> Connect
+sg2 csd << /receptor_type 1 >> Connect
+sg3 csd << /receptor_type 2 >> Connect
+
+100. Simulate
+
+%sd GetStatus /eventinfo
```
Comment (jakobj) [inline examples/nest/correlospinmatrix_detector.sli]: if these lines are not needed (they are commented out) i would remove them from the example.

File: models/correlospinmatrix_detector.cpp [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+    reset = true;
+  }
+
+  if ( updateValue< double_t >( d, names::tau_max, t ) )
+  {
+    tau_max_ = Time::ms( t );
+    reset = true;
+  }
+
+  if ( updateValue< double_t >( d, names::Tstart, t ) )
```
Comment (jakobj) [inline models/correlospinmatrix_detector.cpp]: do we want Tstart and Tstop to be non-negative and also Tstop > Tstart? if so we need checks here.

File: models/correlospinmatrix_detector.h [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+   The correlospinmatrix_detector has a variable number of inputs which can be set via SetStatus
+   under the key N_channels. All incoming connections to a specified receptor will be pooled.
+
+   Parameters:
+   Tstart     double    - Time when to start counting events. This time should be set to at least
+                          start + tau_max in order to avoid edge effects of the correlation counts.
+   Tstop      double    - Time when to stop counting events. This time should be set to at most
+                          Tsim - tau_max, where Tsim is the duration of simulation,
+                          in order to avoid edge effects of the correlation counts.
+   delta_tau  double    - bin width in ms. This has to be an odd multiple of the resolution, to
```
Comment (jakobj) [inline models/correlospinmatrix_detector.h]: the statement that delta_tau has to be a multiple of the resolution seems not to be checked, and the example also explicitly uses an even multiple of the resolution. is this really necessary to mention here? it also seems to work otherwise.

File: models/correlospinmatrix_detector.h [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+   tau_max    double    - one-sided width in ms. In the lower triangular part events with
+   differences in
+                          [0, tau_max+delta_tau/2) are counted. On the diagonal and in
+                          the upper triangular part events with differences in (0,
+   tau_max+delta_tau/2]
+
+   N_channels long      - The number of inputs to correlate. This defines the range of
+   receptor_type. Default is
+   1.
+                          Setting N_channels clears count_covariance, covariance and n_events.
```
Comment (jakobj) [inline models/correlospinmatrix_detector.h]: does changing other parameters also clear count_covariance, etc.? should this not also happen for tau_max and delta_tau?

**Comment by jakobj** [comment]
i added my comment directly to the corresponding lines in the code. i hope this works. in addition i was wondering about tests? do we require them for every model? if so, i think a test is missing for this particular one.

File: models/correlospinmatrix_detector.cpp [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+          long_t l =
+            std::min( t_i_off, t_j_off - Delta * dt ) - std::max( t_i_on, t_j_on - Delta * dt );
+          if ( l > 0 )
+          {
+            S_.count_covariance_[ i ][ j ][ t0 - Delta ] += l;
+            S_.count_covariance_[ j ][ i ][ t0 + Delta ] += l;
+          }
+        }
+
+        if ( i != j )
```
Comment (tammoippen) [inline models/correlospinmatrix_detector.cpp]: Please, always use braces around blocks of code in control structures ([Coding Guidelines](https://nest.github.io/nest-simulator/coding_guidelines_c++#further-indentation-and-formatting)), especially, if they are as nested as here.

File: examples/nest/correlospinmatrix_detector.sli (hunk 1 above)
Comment (mhelias) [inline examples/nest/correlospinmatrix_detector.sli]: done

File: examples/nest/correlospinmatrix_detector.sli (hunk 2 above)
Comment (mhelias) [inline examples/nest/correlospinmatrix_detector.sli]: done

File: models/correlospinmatrix_detector.cpp (hunk 3 above)
Comment (mhelias) [inline models/correlospinmatrix_detector.cpp]: done

File: models/correlospinmatrix_detector.h (hunk 4 above)
Comment (mhelias) [inline models/correlospinmatrix_detector.h]: You are right, this restriction is not existing in this code. Outdated documentation. Updated.

File: models/correlospinmatrix_detector.h (hunk 5 above)
Comment (mhelias) [inline models/correlospinmatrix_detector.h]: you are right, setting the of the reproperties also clears the measurements; updated doc.

File: models/correlospinmatrix_detector.cpp (hunk 6 above)
Comment (mhelias) [inline models/correlospinmatrix_detector.cpp]: thanks for the note, did this.

Comment (mhelias) [inline models/correlospinmatrix_detector.cpp]: added test_correlospinmatrix_detector.sli
and pynest/examples/correlospinmatrix_detector_two_neuron.py

**Comment by mhelias** [comment]
@jakobj @tammoippen 
I'm done now with the points, please have a look.

File: models/correlospinmatrix_detector.cpp [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+      N_channels_ = N;
+      reset = true;
+    }
+  }
+
+  if ( updateValue< double_t >( d, names::delta_tau, t ) )
+  {
+    delta_tau_ = Time::ms( t );
+    reset = true;
+    if ( t < 0 )
```
Comment (jakobj) [inline models/correlospinmatrix_detector.cpp]: shouldn't there be parenthesis around the `throw BadProperty`? also see below.

File: sli/tarrayobj.cc [hunk 8]
```diff
@@ -303,7 +303,7 @@ TokenArrayObj::reduce( Token* first, Token* last )
```
Comment (jakobj) [inline sli/tarrayobj.cc]: i think these changes should not be included in this PR.

File: topology/mask.h [hunk 9]
```diff
@@ -350,7 +350,7 @@ class IntersectionMask : public Mask< D >
   Mask< D >* clone() const;
```
Comment (jakobj) [inline topology/mask.h]: also not these. or am I missing something obvious?

File: sli/fdstreamtest.cc [hunk 10]
```diff
@@ -67,7 +67,8 @@ main()
   }
```
Comment (jakobj) [inline sli/fdstreamtest.cc]: also not these.

File: topology/parameter.h [hunk 11]
```diff
@@ -726,7 +726,7 @@ class ProductParameter : public Parameter
   }
```
Comment (jakobj) [inline topology/parameter.h]: same here.

File: pynest/examples/correlospinmatrix_detector_two_neuron.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+It measures and plots the auto- and cross covariance funcitons
+of the individual neurons and between them, repsectively. 
+
+'''
+
+
+import pylab as pl
+import nest
+import numpy as np
+import scipy.optimize as opt
```
Comment (jakobj) [inline pynest/examples/correlospinmatrix_detector_two_neuron.py]: unused import.

**Comment by mhelias** [comment]
@jakobj corrected white spaces included by automated formatting, hope that worked, otherwise would need help by @tammoIppen to fix this

**Comment by jakobj** [comment]
its seems like it worked. :) i'm happy with this now. :+1:

**Comment by mhelias** [comment]
@tammoippen @lekshmideepu can you have a look at the error? seems to be related to the formatting, for which I used the formatallc++files skript.

**Comment by mhelias** [comment]
@tammoippen now we found the white spaces. Could you now supply the +1, as you told me and cause the thing to be merged?
Many thanks again,
Moritz

**Comment by abigailm** [comment]
No need to wait for Tammo :) :+1:
